br/pkg/lightning/backend/external/util.go

```diff
@@ -42,68 +44,105 @@ func seekPropsOffsets(
 	starts []kv.Key,
 	paths []string,
 	exStorage storage.ExternalStorage,
-	checkHotSpot bool,
 ) (_ [][]uint64, err error) {
 	logger := logutil.Logger(ctx)
 	task := log.BeginTask(logger, "seek props offsets")
 	defer func() {
 		task.End(zapcore.ErrorLevel, err)
 	}()
 
-	// adapt the NewMergePropIter argument types
-	multiFileStat := MultipleFilesStat{Filenames: make([][2]string, 0, len(paths))}
-	for _, path := range paths {
-		multiFileStat.Filenames = append(multiFileStat.Filenames, [2]string{"", path})
+	offsetsPerFile := make([][]uint64, len(paths))
+	for i := range offsetsPerFile {
+		offsetsPerFile[i] = make([]uint64, len(starts))
 	}
-	iter, err := NewMergePropIter(ctx, []MultipleFilesStat{multiFileStat}, exStorage, checkHotSpot)
-	if err != nil {
+	// if **all** files' first key is smaller than start key, we may lost the data
+	// between [start key, min(first key of all files)). So record this case for each
+	// file and check all files afterward.
+	firstKeyTooSmallCheckers := make([]kv.Key, len(paths))
+	eg, egCtx := util.NewErrorGroupWithRecoverWithCtx(ctx)
+	for i := range paths {
+		i := i
+		eg.Go(func() error {
+			r, err2 := newStatsReader(egCtx, exStorage, paths[i], 250*1024)
+			if err2 != nil {
+				if err2 == io.EOF {
+					return nil
+				}
+				return errors.Trace(err2)
+			}
+			defer r.Close()
+
+			moved := false
+			keyIdx := 0
+			curKey := starts[keyIdx]
+
+			p, err3 := r.nextProp()
+			for {
+				switch err3 {
+				case nil:
+				case io.EOF:
+					for keyIdx++; keyIdx < len(starts); keyIdx++ {
+						offsetsPerFile[i][keyIdx] = offsetsPerFile[i][keyIdx-1]
+					}
```

Comment:
```suggestion
					currFileoffset := offsetsPerFile[i][keyIdx]
					for keyIdx++; keyIdx < len(starts); keyIdx++ {
						offsetsPerFile[i][keyIdx] = currFileoffset
					}
```
i think this more clear